clang/lib/DPCT/UserDefinedRules/UserDefinedRules.cpp

```diff
@@ -344,17 +344,141 @@ MetaRuleObject::PatternRewriter::PatternRewriter(
   Subrules = S;
 }
 
+// Read a YAML file recursively and substitute any "!include <filename>"
+// directive with the contents of the referenced file.
+std::unique_ptr<llvm::MemoryBuffer>
+readYAMLFile(const llvm::StringRef &RuleFilePath) {
+  // Load the rule file into a MemoryBuffer
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer =
+      llvm::MemoryBuffer::getFile(RuleFilePath);
+  if (!Buffer) {
+    llvm::errs() << "Error: failed to read " << RuleFilePath << ": "
+                 << Buffer.getError().message() << "\n";
+    clang::dpct::ShowStatus(MigrationErrorInvalidRuleFilePath);
+    dpctExit(MigrationErrorInvalidRuleFilePath);
+  }
+
+  // Get the directory path of the rule file.
+  llvm::SmallString<128> DirectoryPath(RuleFilePath);
+  llvm::sys::path::remove_filename(DirectoryPath);
+
+  // Iterate over the input line by line.
+  std::stringstream Output, IncRuleFilePath;
+  const std::string IncDirective = "!include";
+
+  size_t Ind = 0;
+  bool IncDirectiveFound = false;
+  bool SkipWhiteSpaces = false;
+  bool IsLineBeginning = true;
+
+  llvm::StringRef BuffContent = std::move(*Buffer)->getBuffer();
+  const auto BuffSize = BuffContent.size();
+  while (Ind < BuffSize) {
+    unsigned char ch = BuffContent[Ind];
+
+    // Skip white spaces at the beginning of the line if it contains a directive
+    if (IsLineBeginning) {
+      auto i = Ind;
+      auto c = ch;
+
```

Comment:
Should be camel style, better to use Idx, Ch instead.

Reply:
Done.